### ISSUE TITLE
What are the important metrics that help gauge accuracy and validity of training?

### ISSUE THREAD

**Issue by RageItalic** [huggingface/trl#3893]
Hi, newbie to RL here and just looking to get a deeper understand of the training process and what to look for to ensure that training is going as expected. 

I've been working on an RL based project that uses the GRPO policy and during training, I see an output with a huge table with lots of number including the loss values, reward and reward_std along with KL and lots of other values. 

From what I can understand, since RL is not supervised learning, the loss values are not the most accurate indicators of learning. Is that assumption of mine correct?

Assuming loss values are not the north star metric to look for, what is? What values should I be giving more importance to and are there any trends in the numbers I should be keeping an eye out for (like decreasing loss values are an indicator of proper training in supervised learning)?

I've been looking at the following example notebook from unsloth that fine-tuned a mistral model using the GRPO policy. [LINK](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_v0.3_(7B)-GRPO.ipynb#scrollTo=xbko06oU5Nmb)

I made charts for a few of the columns that are output during the training process and the entire table along with the charts is available [here](https://docs.google.com/spreadsheets/d/1G3Eu5smRRv9jmNZTKpmFG7aNwD4XNztUFc1DTBrrA-8/edit?usp=sharing). 

I'd greatly appreciate any explanations and insights into what trends exist in this data that I may not be seeing.

Thank you so much!

**Comment by mitchelldehaven** [comment]
So I'm fairly new to this as well, but I think you want to be looking `rewards / correctness_reward_func`. This is the primary reward function, determining the correctness of the output relative to gold label. Opposed to normal training where you want loss to be falling as training progresses, you want your reward value to increase. 

There are some additional reward function in your setup, but it is probably useful to think of the correct_reward_func as the primary loss, with the formatting and other reward functions as auxiliary losses meant to steer the model in the right direction toward improving the primary loss function.

Here is a graph from a recent run of mine on GRPO with Qwen3-1.7B. The correctness reward should be increasing as training progresses.

<img width="830" height="493" alt="Image" src="https://github.com/user-attachments/assets/9abe216b-122d-44f6-b522-6c2f969b89df" />

**Comment by RageItalic** [comment]
Thank you so much for this insight! I'm re-running my model training so I can verify this for my own project. But, going back to the example notebook I linked above, I graphed their rewards/correctness_reward_func and the following is what I see:

<img width="881" height="383" alt="Image" src="https://github.com/user-attachments/assets/68ce56c1-527c-4435-8328-9cac67feb0b9" />

Most of the values in the table are 0, but the graph shows a very slight line trending upwards slowly. Is this a good enough indicator of learning? Just want to confirm I understand this correctly. Thanks again!

**Comment by RageItalic** [comment]
Another quick thing, currently training my model, but the table that's being output only has the following columns:

**Step 
Training Loss 
reward 
reward_std 
completions / mean_length 
completions / min_length 
completions / max_length 
completions / clipped_ratio 
completions / mean_terminated_length 
completions / min_terminated_length  
completions / max_terminated_length 
kl 
rewards / reward_fn / mean
rewards / reward_fn / std**

Since there is no "rewards/correctness_reward_func", what value should I be looking at instead?

**Comment by mitchelldehaven** [comment]
A few thoughts here:

With 4 bit, you may run into issues. I'm not super familiar with all this, but 4 bit quantization may degrade performance on math tasks more than general tasks. It may be useful to go up to 8 bits.

Formatting wise, I started from the same notebook and the formatting reward function does not always work properly. For example:
```
The cost of Joe's new HVAC system is $20,000.  It includes 2 conditioning zones, each with 5 vents.  In dollars, what is the cost of the system per vent? 
Answer:
2,000 
Response:
<reasoning>
To find the cost per vent, we first need to determine the total number of vents in the HVAC system. In this case, there are 2 conditioning zones, each with 5 vents, so the total number of vents is 2 zones * 5 vents per zone = 10 vents.

To find the cost per vent, we then divide the total cost of the system ($20,000) by the total number of vents (10): $20,000 / 10 vents = $2,000 per vent.
</reasoning>

<answer>
$2,000 per vent.
</answer> 
Extracted:
$2,000 per vent.
```
Even though this matches the format this reward function won't give it credit:
```
def soft_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]
```
Why? Because if there are any new lines in the `<reasoning>` tags, it won't match, because `.*?` won't match new lines. You will need to change the match function to include `re.DOTALL`


Finally, the actual nature of the problem can be seen from from the above example. 
```
<answer>
$2,000 per vent.
</answer> 
Extracted:
$2,000 per vent.
```
"$2,000 per vent." != "2,000" so no reward is given. The int reward function is aimed at ensure the response is actually a number, so that evaluating result is simple. Regardless, I'm not enough of an expert in terms of reward functions to mitigate this behavior, but frankly you might see some improvement by just modifying the prompt slight to:
```
SYSTEM_PROMPT = """
Respond in the following format. All reasoning should be within the `<reasoning>` tags and the `<answer>` tag should only include the final answer, with no units:
<reasoning>
...
</reasoning>
<answer>
...
</answer>
"""
```

**Comment by RageItalic** [comment]
So, Im just using the notebook example since I looked at it while building my own training pipeline. I'm actually working on an NLP problem around ambiguity detection in questions, not math. So the table headers I sent in the previous comment are what I see from my own training run where I built a custom reward function. That's where I dont see **rewards/correctness_reward_func** as a table column so Im wondering what metric is the stand-in that I can use to see learning happening as expected.

**Comment by mitchelldehaven** [comment]
Whatever your reward function is you should be watching. If you have a custom reward function that should be what you are monitoring for improvement.

**Comment by RageItalic** [comment]
Oh I get it! The correctness_reward _func is the name of a specific reward function in the code itself! I thought it was an objective column during training that should always be there! Makes sense now! 

I need to keep an eye on:

rewards / reward_fn / mean
rewards / reward_fn / std

since my reward function is named reward_fn and I'm assuming these values should be increasing, as you pointed out earlier.

**Comment by qgallouedec** [comment]
Hey @RageItalic, sorry for the slow reply here; we're catching up on the backlog.

Your final take is exactly right: in GRPO the training loss is a surrogate objective and its magnitude/trend isn't a reliable "north star". The reward is.

TRL logs a per-step `reward`/`reward_std` (the weighted sum across all your reward functions) plus one `rewards/<name>/mean` and `rewards/<name>/std` pair per reward function. So `correctness_reward_func` in the unsloth notebook is just the name of that notebook's reward fn; since yours is called `reward_fn`, you watch `rewards/reward_fn/mean` and want it trending up

Thanks to @mitchelldehaven for the great help in the thread!
